File: system/multitier/backend-typescript/src/api/controller/coupon.controller.ts
import { Body, Controller, Get, HttpCode, Post } from '@nestjs/common';
import { CouponService } from '../../core/services/coupon.service';
import { PublishCouponRequest } from '../../core/dtos/publish-coupon-request.dto';
import {
  BrowseCouponsItemResponse,
  BrowseCouponsResponse,
} from '../../core/dtos/browse-coupons-response.dto';

@Controller('api/coupons')
export class CouponController {
  constructor(private readonly couponService: CouponService) {}

  @Post()
  @HttpCode(204)
  async createCoupon(@Body() request: PublishCouponRequest): Promise<void> {
    await this.couponService.createCoupon(
      request.code,
      request.discountRate,
      request.validFrom,
      request.validTo,
      request.usageLimit,
    );
  }

  @Get()
  async browseCoupons(): Promise<BrowseCouponsResponse> {
    const coupons = await this.couponService.getAllCoupons();
    const items: BrowseCouponsItemResponse[] = coupons.map((c) => ({
      code: c.code,
      discountRate: Number(c.discountRate),
      validFrom: c.validFrom ? new Date(c.validFrom).toISOString() : null,
      validTo: c.validTo ? new Date(c.validTo).toISOString() : null,
      usageLimit: c.usageLimit,
      usedCount: c.usedCount,
    }));
    return { coupons: items };
  }
}

File: system/multitier/backend-dotnet/Controllers/CouponController.cs
using Microsoft.AspNetCore.Mvc;
using Optivem.Shop.Backend.Core.Dtos;
using Optivem.Shop.Backend.Core.Services;

namespace Optivem.Shop.Backend.Controllers;

[ApiController]
[Route("api/coupons")]
public class CouponController : ControllerBase
{
    private readonly CouponService _couponService;

    public CouponController(CouponService couponService)
    {
        _couponService = couponService;
    }

    [HttpPost]
    public async Task<IActionResult> CreateCoupon([FromBody] PublishCouponRequest request)
    {
        await _couponService.CreateCouponAsync(
            request.Code!,
            request.DiscountRate!.Value,
            request.ValidFrom,
            request.ValidTo,
            request.UsageLimit
        );
        return NoContent();
    }

    [HttpGet]
    public async Task<IActionResult> BrowseCoupons()
    {
        var coupons = await _couponService.GetAllCouponsAsync();
        var items = coupons.Select(c => new BrowseCouponsItemResponse
        {
            Code = c.Code,
            DiscountRate = c.DiscountRate,
            ValidFrom = c.ValidFrom,
            ValidTo = c.ValidTo,
            UsageLimit = c.UsageLimit,
            UsedCount = c.UsedCount
        }).ToList();

        return Ok(new BrowseCouponsResponse { Coupons = items });
    }
}
